File: Euler-Particles/src/shape/ShapeFactory.java
package shape;

import processing.core.PApplet;

public class ShapeFactory {
    private final int CIRCLE = 0;
    private final int POLYGON = 1;
    private final int STAR = 2;
    private final int RANDOMPOLYGONS = 3;
    private final int RANDOMSTARS = 4;
    private PApplet p;
    private int type;
    private int sides;

    public ShapeFactory(PApplet p) {
        this.p = p;
    }

    public Shape getShape() {
        Shape shape;

        switch (type) {
            case POLYGON:
                shape = new Polygon(p, sides);
                break;
            case STAR:
                shape = new Star(p, sides);
                break;
            case RANDOMPOLYGONS:
                shape = new Polygon(p);
                break;
            case RANDOMSTARS:
                shape = new Star(p);
                break;
            default:
                shape = new Circle(p);
                break;
        }

        return shape;
    }

    public void randomize() {
        int r = (int)p.random(500);

        switch (r % 5) {
            case 1:
                type = POLYGON;
                break;
            case 2:
                type = STAR;
                break;
            case 3:
                type = RANDOMPOLYGONS;
                break;
            case 4:
                type = RANDOMSTARS;
                break;
            default:
                type = CIRCLE;
                break;
        }

        sides = (int)p.random(3, 12);
    }
}

File: docs/Euler-Particles/shape/shape_factory.js
class ShapeFactory {

    constructor() {
        this._CIRCLE = 0;
        this._POLYGON = 1;
        this._STAR = 2;
        this._RANDOMPOLYGON = 3;
        this._RANDOMSTAR = 4;
        this.type = this._CIRCLE;
        this.sides = 0;
        this.randomize();
    }

    getShape() {
        let shape = null;

        switch (this.type) {
            case this._POLYGON:
                shape = new Polygon(this.sides);
                break;
            case this._STAR:
                shape = new Star(this.sides);
                break;
            case this._RANDOMPOLYGON:
                shape = new Polygon(int(random(3, 12)));
                break;
            case this._RANDOMSTAR:
                shape = new Star(int(random(4, 12)));
                break;
            default:
                shape = new Circle();
                break;
        }

        return shape;
    }

    randomize() {
        let r = int(random(500));

        switch (r % 5) {
            case 1:
                this.type = this._CIRCLE;
                break;
            case 2:
                this.type = this._POLYGON;
                break;
            case 3:
                this.type = this._RANDOMPOLYGON;
                break;
            case 4:
                this.type = this._RANDOMSTAR;
                break;
            default:
                this.type = this._STAR;
                break;
        }

        this.sides = int(random(3, 12));
    }
}
